snakemake_storage_plugin_xrootd/__init__.py

```diff
@@ -16,178 +21,170 @@
 from snakemake_interface_storage_plugins.storage_object import (
     StorageObjectRead,
     StorageObjectWrite,
-    StorageObjectGlob,
     retry_decorator,
 )
 from snakemake_interface_storage_plugins.io import IOCacheStorageInterface
 
 
-def parse_query_params(query_params: List[str]):
-    """Parse query params from command line."""
-    return dict(param.split("=") for param in query_params)
+class XRootDFatalException(Exception):
+    """
+    Used to prevent retries for certain XRootD errors with the retry decorator
+    """
+
+    pass
 
 
-def unparse_query_params(query_params: Mapping[str, str]):
-    """Unparse query params to command line."""
-    return [f"{key}={value}" for key, value in query_params.items()]
+def _raise_fatal_error(exception: Type[Exception]):
+    if isinstance(exception, XRootDFatalException):
+        get_logger().warning("Unrecoverable error, no more retries")
+        raise exception
+
+
+# TODO configurable would be nice
+xrootd_retry = retry(
+    tries=3, delay=3, backoff=2, logger=get_logger(), fail_callback=_raise_fatal_error
+)
 
 
 @dataclass
 class StorageProviderSettings(StorageProviderSettingsBase):
-    host: Optional[str] = field(
```

Comment:
Why is host, port, username, password removed?

Reply:
So for me I'm not sure if this is just the way I use XRootD but I was not even aware the username and password could be used in the URI like this to do the authentication.

Typically, the authentication (in my experience) is done "in the background" by looking for a valid kerberos token and/or a VOMS proxy certificate.

Actually the other authentication type I have seen is using a token on the end of the root URI but that was typically done for each file separately so not sure if there is much point adding that to the `StorageProvider` itself.

For the host this was something I wondered about a bit. I ended up removing it from the `StorageProvider` because *usually* when I use XRootD with snakemake I get a fully-formed root URI from somewhere else that can then just be plugged straight in so it would avoid an extra configuration step in snakemake itself

Reply:
I see, makes sense.